text/0006-shipyard-next-generation.md

```diff
@@ -0,0 +1,317 @@
+
+### Keptn CloudEvents
+
+Based on the propsed concepts, changes on the Keptn CloudEvents are necessary.
+
+**(1) Reserved space in CloudEvent payload:**
+
+Before the shipyard controller sends out a `[task].triggered` event for a specific task, the controller adds a reserved space to the event payload. This reserved space is named after the task and can be consumed by the unit (Keptn-service) that executes the task. For example, a *deployment* task gets the reserved space of *deployment* in the payload (i.e., data field) of the CloudEvent. Per default, this reserved space is empty. Expect for tasks that have properties defined in the shipyard (e.g., `strategy: blue_green`) - for those tasks the properties are added to the reserved space:
+  
+```
+{
+  "type": "sh.keptn.event.deployment.triggered",
+  ...
+  "data": {
+    "deployment": {
+      "strategy": "blue_green",
+      "traffic": "100",
+    }
+  }
+}
+```
+
+*Implementation details:* 
+* Keptn-services can add any information to the reserved space. For example:
+  ```
+  {
+    "type": "sh.keptn.event.deployment.triggered",
+    ...
+    "data": {
+      "deployment": {
+        "strategy": "blue_green",
+        "traffic": "100",
+        "deploymentURI":"https://my-service.domain.com/" // Property added by Keptn-service
+      }
+    }
+  }
+  ```
+
+* The event selectors for Keptn-services, as proposed in [First version of a Keptn uniform](https://github.com/keptn/enhancement-proposals/pull/7), work based on the information provided in this reserved space. For example:
+  ```yaml
+    - name: deployment-service
+      image: keptn/helm-service:0.6.0
+      events:
+      - deployment.triggered:
+        selector:
+          matchExpressions:
+          - {key: strategy, operator: In, values: [direct, blue_green]}
+  ```
+
+* The shipyard controller merges the reserved spaces of a `[task-xyz].triggered`, `[task-xyz].started` and `[task-xyz].finished` event. If, for example, the `deployment.triggered` event contains `strategy:blue_green` and the `deployment.finished` event contains `deploymentURI:xyz` in the reserved space, the shipyard controller merges the the reserved space, before sending the next `[test].triggered`: 
+  ```
+  {
+    "type": "sh.keptn.event.task-abc.triggered",
+    ...
+    "data": {
+      "deployment": {
+        "strategy": "blue_green",
+        "traffic": "100",
+        "deploymentURI":"https://my-service.domain.com/" // Property added by Keptn-service
+      }
+      "test": {
+
+      }
+    }
+  }
+  ```
+
+
+
+## Breaking changes
+
+This KEP breaks the implementation of Keptn 0.6.0: 
+* Keptn-services (aka. batteries) including contributed services need to be changed to react on a `[task].triggered`, and send a `[task].started` → `[task].finished`.
+* Keptn CloudEvents must be changed.
+* A shipyard migration must take place as outlined below.
+
+### Upgrade path from Shipyard 0.1.2  to 0.2.0
```

Comment:
Let's not use 0.1.2, but *0.1.x* when we refer to the old Shipyard within this KEP